microprofile/microprofile-fault-tolerance-2.0/src/test/java/org/wildfly/swarm/ts/microprofile/fault/tolerance/v20/FaultTolerance20AsyncPartialTest.java

```diff
@@ -0,0 +1,177 @@
+        expectedResponses.put("Hello from @Bulkhead @Timeout method", new IntRange(0, 20));
+        expectedResponses.put("Fallback Hello", new IntRange(20, 40));
+        testPartial(40, "http://localhost:8080/partial?operation=bulkhead15q5-timeout&counter=", expectedResponses);
+    }
+
+    /**
+     * Test sends 40 parallel requests. There are annotations on service:
+     * timeout 1s and bulkhead (of maximum number of concurrent calls = 5 and 5 queued).
+     * Every even call is made to timeout. Calls above 20 fall-back.
+     * Tests expects maximal success of 20 messages (those not timeout-ed + not fallback-ed).
+     */
+    @Test
+    @RunAsClient
+    public void bulkhead5q5Timeout() throws InterruptedException {
+        Map<String, IntRange> expectedResponses = new HashMap<>();
+        expectedResponses.put("Hello from @Bulkhead @Timeout method", new IntRange(0, 20));
+        expectedResponses.put("Fallback Hello", new IntRange(20, 40));
+        testPartial(40, "http://localhost:8080/partial?operation=bulkhead5q5-timeout&counter=", expectedResponses);
+    }
+
+    /**
+     * Test sends 20 parallel requests. There are annotations on service:
+     * timeout 1s, bulkhead (of maximum number of concurrent calls = 5 and 5 queued), retry
+     * 5 requests invoke timeout exception immediately, 5 requests time-out. Calls above 10 fall-back.
+     * Tests expects maximal success of 10 messages (those without exception, not timeout-ed + not fallback-ed).
+     */
+    @Test
+    @RunAsClient
+    public void bulkhead5q5TimeoutRetry() throws InterruptedException {
+        Map<String, IntRange> expectedResponses = new HashMap<>();
+        expectedResponses.put("Hello from @Bulkhead @Timeout @Retry method", new IntRange(0, 10));
+        expectedResponses.put("Fallback Hello", new IntRange(10, 20));
+        testPartial(20, "http://localhost:8080/partial?operation=bulkhead5q5-timeout-retry&counter=", expectedResponses);
+    }
+
+    /**
+     * Test sends 20 parallel requests. There are annotations on service:
+     * timeout 1s, bulkhead (of maximum number of concurrent calls = 15 and 5 queued), retry
+     * 5 requests invoke timeout exception immediately, 5 requests time-out.
+     * Tests expects maximal success of 10 messages (those without exception, not timeout-ed).
+     */
+    @Test
+    @RunAsClient
+    public void bulkhead15q5TimeoutRetry() throws InterruptedException {
+        Map<String, IntRange> expectedResponses = new HashMap<>();
+        expectedResponses.put("Hello from @Bulkhead @Timeout @Retry method", new IntRange(0, 10));
+        expectedResponses.put("Fallback Hello", new IntRange(10, 20));
+        testPartial(20, "http://localhost:8080/partial?operation=bulkhead15q5-timeout-retry&counter=", expectedResponses);
+    }
+
+    private static void testPartial(int parallelRequests, String url, Map<String, IntRange> expectedResponses) throws InterruptedException {
```

Comment:
Method should be at the end of the class. Also I don't like `IntRange`, from `new IntRange(0, 10)` it's not really clear if `0` or `10` are included or not. I'm not sure if we already have a dependency on Guava, but if we do, I'd always prefer Guava over Apache Commons.

Reply:
Range.closed or Range.openClosed looks better, modified to Guava